driver/src/price_estimation/mod.rs

```diff
@@ -1,30 +1,151 @@
 //! Module responsible for aggregating price estimates from various sources to
 //! give good price estimates to the solver for better results.
 
-#[allow(dead_code)]
 mod kraken;
 
-use crate::models::{TokenId, TokenInfo};
+use self::kraken::KrakenClient;
+use crate::models::{Order, TokenData, TokenId, TokenInfo};
 use anyhow::Result;
-use ethcontract::H160;
-use std::collections::HashMap;
+use lazy_static::lazy_static;
+use log::warn;
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::iter;
+
+/// A type alias for token information map that is passed to the solver.
+type Tokens = BTreeMap<TokenId, Option<TokenInfo>>;
+
+/// A trait representing a price oracle that retrieves price estimates for the
+/// tokens included in the current orderbook.
+#[cfg_attr(test, mockall::automock)]
+pub trait PriceEstimating {
+    fn get_token_prices(&self, orders: &[Order]) -> Tokens;
+}
+
+pub struct PriceOracle {
+    /// The token data supplied by the environment. This ensures that only
+    /// whitelisted tokens get their prices estimated.
+    tokens: TokenData,
+    /// The price source to use.
+    ///
+    /// Note that currently only one price source is supported, but more price
+    /// source are expected to be added in the future.
+    source: Box<dyn PriceSource>,
+}
+
+impl PriceOracle {
+    /// Creates a new price oracle from a token whitelist data.
+    pub fn new(tokens: TokenData) -> Result<Self> {
+        Ok(PriceOracle::with_source(tokens, KrakenClient::new()?))
+    }
+
+    fn with_source(tokens: TokenData, source: impl PriceSource + 'static) -> Self {
+        PriceOracle {
+            tokens,
+            source: Box::new(source),
+        }
+    }
+
+    /// Splits order tokens into a vector of tokens that should be priced based
+    /// on the token whitelist and a vector of unpriced token ids.
+    ///
+    /// Note that all token ids in the returned results are garanteed to be
+    /// unique.
+    fn split_order_tokens(&self, orders: &[Order]) -> (Vec<Token>, Vec<TokenId>) {
+        let unique_token_ids: HashSet<_> = orders
+            .iter()
+            .flat_map(|order| vec![order.buy_token, order.sell_token])
+            .map(TokenId)
+            // NOTE: Always include the reference token. This is done since the
+            //   solver input specifies the reference token, so for correctness
+            //   it should always be considered.
+            .chain(iter::once(TokenId::reference()))
+            .collect();
+
+        unique_token_ids.into_iter().fold(
```

Comment:
Why not https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition?

Reply:
Actually never mind. Saw you response here: https://github.com/gnosis/dex-services/pull/578#discussion_r383028252